src/classes/Collision/CapsuleConvexHull.js

```diff
@@ -0,0 +1,240 @@
+            innerSegmentEnd,
+            referenceFace,
+            clippedInnerSegmentStart,
+            clippedInnerSegmentEnd
+        );
+        if ( !clippingWasSuccessful ) {
+            Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contact, closestPointOnInnerSegment, closestPointOnHull );
+            return [ contact ];
+        }
+
+        var contactAtTheStart = contact;
+        var contactAtTheEnd = Goblin.ObjectPool.getObject( 'ContactDetails' );
+        contactAtTheEnd.object_a = capsule;
+        contactAtTheEnd.object_b = convexHull;
+        contactAtTheEnd.contact_normal.copy( contactAtTheStart.contact_normal );
+        contactAtTheEnd.penetration_depth = contactAtTheStart.penetration_depth;
+
+        contactAtTheStart.contact_point_in_b.scaleVector( contactNormal, capsule.shape.radius - penetrationDepth );
+        contactAtTheStart.contact_point_in_b.add( clippedInnerSegmentStart );
+        contactAtTheEnd.contact_point_in_b.scaleVector( contactNormal, capsule.shape.radius - penetrationDepth );
+        contactAtTheEnd.contact_point_in_b.add( clippedInnerSegmentEnd );
+
+        Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contactAtTheStart, clippedInnerSegmentStart, contactAtTheStart.contact_point_in_b );
+        Goblin.Collision.capsuleConvexHull._fillContactWithDetails( contactAtTheEnd, clippedInnerSegmentEnd, contactAtTheEnd.contact_point_in_b );
+
+        return [ contactAtTheStart, contactAtTheEnd ];
+    };
+} )();
+
+Goblin.Collision.capsuleConvexHull._deepCapsuleConvexHull = ( function() {
+    return function( capsule, convexHull, doLightweightCollision ) {
+        /**
+         * @type {Goblin.ContactDetails}
+         */
+        var contact = Goblin.ObjectPool.getObject( 'ContactDetails' );
+        contact.object_a = capsule;
+        contact.object_b = convexHull;
+        if ( doLightweightCollision ) {
+            contact.is_lightweight = true;
+            return [ contact ];
+        }
+
+        return null;
+    };
+} )();
+
+/**
+ *
+ * @param {Goblin.ContactDetails} contact
+ * @param {Goblin.Vector3} closestPointOnInnerSegment - closest point on inner segment for the objectA
+ * @param {Goblin.Vector3} closestPointOnHull - closest point on inner segment for the objectB
+ */
+Goblin.Collision.capsuleConvexHull._fillContactWithDetails = function( contact, closestPointOnInnerSegment, closestPointOnHull ) {
+    contact.contact_point_in_a.scaleVector( contact.contact_normal, contact.object_a.shape.radius );
+    contact.contact_point_in_a.add( closestPointOnInnerSegment );
+
+    contact.contact_point_in_b.copy( closestPointOnHull );
+
+    contact.contact_point.addVectors( contact.contact_point_in_a, contact.contact_point_in_b );
+    contact.contact_point.scale( 0.5 );
+
+    contact.object_a.transform_inverse.transformVector3( contact.contact_point_in_a );
+    contact.object_b.transform_inverse.transformVector3( contact.contact_point_in_b );
+};
+
+/**
+ *
```

Comment:
And this one as well - and any other without a description :)

Reply:
It's a little bit early to review this pull request - that's why it's marked as WIP and don't have assigned reviewers 😄   Nevertheless, I will address all your comments

Reply:
fixed